storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2812,6 +2816,74 @@ fseg_free_extent(
 	return DB_SUCCESS;
 }
 
+/** Free the extent and fragment page associated with
+the segment.
+@param iblock     page where segment header are placed
+@param inode      index node information
+@param space      tablespace where segment resides
+@param mtr        mini-transaction
+@param header     segment header
+@param free_inode free the inode page also
+@param ahi        adaptive hash index
+@return DB_SUCCESS_LOCKED_REC when freeing wasn't completed
+@return DB_SUCCESS or other error code  when freeing was completed */
+static
+dberr_t fseg_free_step_low(buf_block_t *iblock, fseg_inode_t *inode,
+                           fil_space_t *space, mtr_t *mtr,
+                           fseg_header_t *header, bool free_inode
+#ifdef BTR_CUR_HASH_ADAPT
+                           , bool ahi=false
+#endif /* BTR_CUR_HASH_ADAPT */
+                           )
+{
+  dberr_t err= DB_SUCCESS;
+  if (xdes_t *descr= fseg_get_first_extent(inode, space, mtr, &err))
+  {
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                          , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                        );
+    return err == DB_SUCCESS ? DB_SUCCESS_LOCKED_REC : err;
+  }
+
+  if (err != DB_SUCCESS)
+    return err;
+
+  /* Free a fragment page. If there are no fragment pages
+  exist in the array then free the file segment inode */
+  ulint n = fseg_find_last_used_frag_page_slot(inode);
+  if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+    return free_inode
+      ? fsp_free_seg_inode(space, inode, iblock, mtr)
+      : DB_SUCCESS;
+
+  /* fseg_free_step_not_header() should not freed the header page
+  fseg_free_step() should free the header page and free
+  the file segment inode. Header page should be the first
+  one to be allocated in fragment array of file segment. */
+  uint32_t page_no = fseg_get_nth_frag_page_no(inode, n);
+  if (header && page_no == page_get_page_no(page_align(header)))
+     return DB_SUCCESS;
+
+  err= fseg_free_page_low(inode, iblock, space, page_no, mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                          , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                          );
+  if (err != DB_SUCCESS)
+    return err;
+  buf_page_free(space, page_no, mtr);
+  if (free_inode)
+  {
+    n= fseg_find_last_used_frag_page_slot(inode);
+    if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+      return fsp_free_seg_inode(space, inode, iblock, mtr);
+  }
+  return DB_SUCCESS_LOCKED_REC;
```

Comment:
I’d simplify this a bit:
```c++
  if (!header_page_no &&
      fseg_find_last_used_frag_page_slot(inode) == ULINT_UNDEFINED)
    return fsp_free_seg_inode(space, inode, iblock, mtr);
  return DB_SUCCESS_LOCKED_REC;
```
Can we define the parameters of `fseg_free_page_low()` in more similar order as `fseg_free_seg_inode()`? That is: `space, inode, iblock, mtr, page_no, ahi`? Likewise, the first parameters of this function `fseg_free_step_low()` should be `space, inode, iblock, header_page_no, ahi`. In that way, there would be less shuffling of parameter registers needed.